sw/device/silicon_creator/lib/ownership/ownership.c

```diff
@@ -1,21 +1,177 @@
+  // For now, we consider the seal to be just the sha256 over the page.
+  // This really needs to be the KMAC over the page with a keymgr-created
+  // sealing secret.
+  hmac_digest_t digest;
+  hmac_sha256(&owner_page[page], seal_len, &digest);
+  if (hardened_memeq(owner_page[page].seal, digest.digest,
+                     ARRAYSIZE(digest.digest)) == kHardenedBoolTrue) {
+    return kOwnerPageStatusSealed;
+  }
+  hardened_bool_t result = ownership_key_validate(
+      /*page=*/page, kOwnershipKeyOwner, &owner_page[page].signature,
+      &owner_page[page], sig_len);
+  if (result == kHardenedBoolFalse) {
+    // If the page is bad, destroy the RAM copy.
+    memset(&owner_page[page], 0x5a, sizeof(owner_page[0]));
+  }
+  return result == kHardenedBoolTrue ? kOwnerPageStatusSigned
+                                     : kOwnerPageStatusInvalid;
+}
 
-rom_error_t ownership_init(void) {
+OT_WEAK rom_error_t test_owner_init(boot_data_t *bootdata,
+                                    owner_config_t *config,
+                                    owner_application_keyring_t *keyring) {
+  dbg_printf("error: no owner.\r\n");
+  return kErrorOwnershipNoOwner;
+}
+
+static rom_error_t locked_owner_init(boot_data_t *bootdata,
+                                     owner_config_t *config,
+                                     owner_application_keyring_t *keyring) {
+  if (owner_page_valid[0] == kOwnerPageStatusInvalid &&
+      owner_page_valid[1] == kOwnerPageStatusSealed) {
+    // Page 0 bad, Page 1 good: copy page 1 to page 0.
+    memcpy(&owner_page[0], &owner_page[1], sizeof(owner_page[0]));
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(
+        &kFlashCtrlInfoPageOwnerSlot0, kFlashCtrlEraseTypePage));
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
+        &kFlashCtrlInfoPageOwnerSlot0, 0,
+        sizeof(owner_page[0]) / sizeof(uint32_t), &owner_page[0]));
+    owner_page_valid[0] = owner_page_valid[1];
+  } else if (owner_page_valid[1] == kOwnerPageStatusInvalid &&
+             owner_page_valid[0] == kOwnerPageStatusSealed) {
+    // Page 1 bad, Page 0 good: copy page 0 to page 1.
+    memcpy(&owner_page[1], &owner_page[0], sizeof(owner_page[0]));
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(
+        &kFlashCtrlInfoPageOwnerSlot1, kFlashCtrlEraseTypePage));
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
+        &kFlashCtrlInfoPageOwnerSlot1, 0,
+        sizeof(owner_page[1]) / sizeof(uint32_t), &owner_page[1]));
+    owner_page_valid[1] = owner_page_valid[0];
+  } else if (owner_page_valid[0] != kOwnerPageStatusSealed &&
+             owner_page_valid[1] != kOwnerPageStatusSealed) {
+    // Neither page is valid; go to the LockedNone state.
+    dbg_printf("error: both owner pages invalid.\r\n");
+    bootdata->ownership_state = kOwnershipStateLockedNone;
+    nonce_new(&bootdata->nonce);
+    HARDENED_RETURN_IF_ERROR(boot_data_write(bootdata));
+    return kErrorOwnershipBadInfoPage;
```

Comment:
What do you think about `s/kOwnershipStateLockedNone/kOwnershipStateRecovery/`.

However, is there a way to recover here? It feels that the only mechanism available would be to bootstrap an image to initialize the chip. We could also include a recovery key in the ROM_EXT to support owner recovery if bootstrap is not available...

Reply:
Renamed `LockedNone` to `Recovery`.

The `silicon_creator` has the option to compile into the ROM_EXT a `kNoOwnerRecoveryKey`.  If that key is present in the ROM_EXT, then that key can be used with the ownership unlock command to transition the chip from `OwnershipStateLockedNone` to one of the unlocked states (`Unlocked{Any, Endorsed}`) and then re-apply an ownership configuration.

If a chip finds its way to the `LockedNone` state, the options then are:
1. Use the No-owner recovery key and the rescue protocol to recover the chip.
2. SPI-bootstrap a special ROM_EXT to forcibly recover the chip.

Both operations involve `silicon_creator` unless `silicon_creator` delegates use of the No-owner recovery key to another entity.